rust/src/operations/optimize.rs

```diff
@@ -641,61 +650,90 @@ impl MergePlan {
                     object_store.clone(),
                     self.max_spill_size,
                 )?);
+                let task_parameters = self.task_parameters.clone();
+                let object_store = object_store.clone();
                 futures::stream::iter(bins)
-                    .map(|(partition, files)| {
+                    .map(move |(partition, files)| {
                         let batch_stream = Self::read_zorder(files.clone(), exec_context.clone());
 
                         let object_store = object_store.clone();
 
                         let rewrite_result = tokio::task::spawn(Self::rewrite_files(
-                            self.task_parameters.clone(),
+                            task_parameters.clone(),
                             partition,
                             files,
                             object_store,
                             batch_stream,
                         ));
                         util::flatten_join_error(rewrite_result)
                     })
-                    .buffer_unordered(self.max_concurrent_tasks)
-                    .try_for_each(|(partial_actions, partial_metrics)| {
-                        debug!("Recording metrics for a completed partition");
-                        actions.extend(partial_actions);
-                        metrics.add(&partial_metrics);
-                        async { Ok(()) }
-                    })
-                    .await?;
+                    .boxed()
             }
-        }
+        };
 
-        metrics.preserve_insertion_order = true;
-        if metrics.num_files_added == 0 {
-            metrics.files_added.min = 0;
-        }
-        if metrics.num_files_removed == 0 {
-            metrics.files_removed.min = 0;
-        }
+        let mut stream = stream.buffer_unordered(self.max_concurrent_tasks);
 
-        // TODO: Check for remove actions on optimized partitions. If a
-        // optimized partition was updated then abort the commit. Requires (#593).
-        if !actions.is_empty() {
-            let mut metadata = Map::new();
-            metadata.insert("readVersion".to_owned(), self.read_table_version.into());
-            let maybe_map_metrics = serde_json::to_value(metrics.clone());
-            if let Ok(map) = maybe_map_metrics {
-                metadata.insert("operationMetrics".to_owned(), map);
+        let mut table = DeltaTable::new_with_state(object_store.clone(), snapshot.clone());
+
+        let mut last_commit = Instant::now();
+        loop {
+            let next = stream.next().await.transpose()?;
```

Comment:
nit: For clarity, let's initialize the variables we are collecting right above the loop where we are collecting them.

```suggestion
        // Actions buffered so far. These will be flushed either at the end
        // or when we reach the commit interval.
        let mut buffered_actions = vec![];
        
        // Each time we commit, we'll reset buffered_metrics to orig_metrics.
        let orig_metrics = std::mem::take(&mut self.metrics);
        let mut buffered_metrics = orig_metrics.clone();
        let mut total_metrics = orig_metrics.clone();
        
        let mut last_commit = Instant::now();
        loop {
            let next = stream.next().await.transpose()?;
```

Reply:
Applied this.